crates/hc_seed_bundle/src/seed_cipher.rs

```diff
@@ -1,20 +1,355 @@
+            passphrase,
+            salt,
+            mem_limit,
+            ops_limit,
+            seed_cipher_header,
+            seed_cipher,
+        )
+        .await?;
+
+        // build the "unlocked" seed bundle struct with the seed
+        let mut bundle =
+            crate::UnlockedSeedBundle::priv_from_seed(seed).await?;
+
+        // apply the app_data
+        bundle.set_app_data_bytes(app_data);
+
+        Ok(bundle)
+    }
+}
+
+/// This locked cipher is based on security questions.
+pub struct LockedSeedCipherSecurityQuestions {
+    salt: sodoken::BufReadSized<16>,
+    mem_limit: usize,
+    ops_limit: u64,
+    question_list: (String, String, String),
+    seed_cipher_header: sodoken::BufReadSized<24>,
+    seed_cipher: sodoken::BufReadSized<49>,
+    app_data: Arc<[u8]>,
+}
+
+impl std::fmt::Debug for LockedSeedCipherSecurityQuestions {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("LockedSeedCipherSecurityQuestions")
+            .field("question_list", &self.question_list)
+            .finish()
+    }
+}
+
+impl LockedSeedCipherSecurityQuestions {
+    /// List the questions
+    pub fn get_question_list(&self) -> &(String, String, String) {
+        &self.question_list
+    }
+
+    /// Decrypt this Cipher into an UnlockedSeedBundle struct.
+    pub async fn unlock<A>(
+        self,
+        answer_list: (A, A, A),
+    ) -> SodokenResult<crate::UnlockedSeedBundle>
+    where
+        A: Into<sodoken::BufRead> + 'static + Send,
+    {
+        // destructure our decoding data
+        let LockedSeedCipherSecurityQuestions {
+            salt,
+            mem_limit,
+            ops_limit,
+            seed_cipher_header,
+            seed_cipher,
+            app_data,
+            ..
+        } = self;
+
+        // generate a deterministic passphrase with the given answers
+        let (a1, a2, a3) = answer_list;
+        let passphrase = process_security_answers(a1, a2, a3)?;
+
+        // decrypt the seed with the generated passphrase
+        let seed = pw_dec(
+            passphrase,
+            salt,
+            mem_limit,
+            ops_limit,
+            seed_cipher_header,
+            seed_cipher,
+        )
+        .await?;
+
+        // build the "unlocked" seed bundle struct with the seed
+        let mut bundle =
+            crate::UnlockedSeedBundle::priv_from_seed(seed).await?;
+
+        // apply the app_data
+        bundle.set_app_data_bytes(app_data);
+
+        Ok(bundle)
+    }
+}
+
+/// Enum of Locked SeedCipher types handled by this library.
+#[non_exhaustive]
+#[derive(Debug)]
+pub enum LockedSeedCipher {
+    /// This locked cipher is a simple pwHash type.
+    PwHash(LockedSeedCipherPwHash),
+
+    /// This locked cipher is based on security question answers.
+    SecurityQuestions(LockedSeedCipherSecurityQuestions),
+
+    /// The type-name of a cipher not yet supported by this library.
+    UnsupportedCipher(Box<str>),
```

Comment:
What's the purpose of this?

Reply:
Forward compatibility... if this version of the lib is reading a cipher produced by a later version of the lib, it can at least say the name of the cipher, even if it can't decrypt it.